{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "PhARIS.ipynb",
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/JKrusche1/PhARIS/blob/main/PhARIS.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "src_JK"
      },
      "source": [
        "<img src=\"https://raw.githubusercontent.com/JKrusche1/PhARIS/main/phage_logo.png\" height=\"200\" align=\"right\" style=\"height:200px\">\n",
        "\n",
        "##**PhARIS - v1.0**\n",
        "##**Ph**age **A**ureus **R**BP **I**dentification **S**ystem\n",
        "\n",
        "Tool for rapid detection of _Staphylococcus aureus_ phage receptor-binding proteins (RBPs) by blast homology search of a query with known _S. aureus_ RBP clusters.\n",
        "\n",
        "Created by Janes Krusche\n",
        "\n",
        "2023 - manuscript in preparation\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title Install dependencies\n",
        "#@markdown #### Install biopython, blast+ and RBP reference files\n",
        "\n",
        "from google.colab import files\n",
        "import io\n",
        "import os\n",
        "import json\n",
        "\n",
        "print(\"Installing biopython...\")\n",
        "!pip install biopython > /dev/null\n",
        "print(\"Done\")\n",
        "print(\"Installing blast...\")\n",
        "!apt-get install ncbi-blast+ > /dev/null\n",
        "print(\"Done\")\n",
        "print(\"Downloading reference files...\")\n",
        "if not os.path.exists(\"files\"):\n",
        "  !git clone https://github.com/JKrusche1/PhARIS.git &>/dev/null\n",
        "  !mv PhARIS/files /content/\n",
        "  !rm -r PhARIS\n",
        "  if not os.path.exists(\"files/blast_main\"):\n",
        "      !mkdir files/blast_main\n",
        "  if not os.path.exists(\"files/blast_rev\"):\n",
        "      !mkdir files/blast_rev\n",
        "  # Read the master dictionary from the JSON file\n",
        "  with open(\"files/RBP.json\", \"r\") as json_file:\n",
        "      master_dict = json.load(json_file)\n",
        "\n",
        "  # Access individual dictionaries using their names\n",
        "  clusterlist_main = master_dict[\"clusterlist_main\"]\n",
        "  clusterlist_rev = master_dict[\"clusterlist_rev\"]\n",
        "  FACS_values = master_dict[\"FACS_values\"]\n",
        "  FACS_SEM = master_dict[\"FACS_SEM\"]\n",
        "  RBP_defi = master_dict[\"RBP_defi\"]\n",
        "print(\"Done\")"
      ],
      "metadata": {
        "id": "FqafU92tRpZF",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "g5kqJMhGkCq5",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title ### Search for S. aureus phage RBP\n",
        "#@markdown Once executed, a button will appear that lets you upload your phage genome in fasta format.\\\n",
        "#@markdown After the upload is finished, the script will search for an _S. aureus_ binding RBP in your genome.\\\n",
        "#@markdown If an RBP is found, you will get the closest matching RBP of the clustering, as well as the identity to this RBP.\\\n",
        "#@markdown To find out where your RBP is located in the phylogenetic tree, go to the links provided and search for the closest match.\\\n",
        "#@markdown [Primary RBP tree](https://itol.embl.de/tree/4652552332411677924980)\\\n",
        "#@markdown [Secondary RBP tree](https://itol.embl.de/tree/46525523312761688382866)\n",
        "\n",
        "#@markdown To search another genome, simply rerun this cell and upload another file.\n",
        "\n",
        "#Import libraries\n",
        "from Bio import SeqIO\n",
        "import os\n",
        "import glob\n",
        "import pathlib\n",
        "import pandas as pd\n",
        "import sys\n",
        "import time\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "\n",
        "#Upload file\n",
        "# Prompt the user to upload a file\n",
        "uploaded = files.upload()\n",
        "\n",
        "# Check if any files were uploaded\n",
        "if uploaded:\n",
        "    # If files were uploaded, execute the following block of code\n",
        "    print('File uploaded successfully.')\n",
        "\n",
        "    # Save the uploaded file to the \"files\" directory with the name \"genome.fasta\"\n",
        "    with open(os.path.join('files', 'genome.fasta'), 'wb') as f:\n",
        "        # Write the contents of the uploaded file to the newly created \"genome.fasta\" file\n",
        "        # The \"next(iter(uploaded))\" extracts the name of the uploaded file\n",
        "        # The \"uploaded[next(iter(uploaded))]\" fetches the contents of the uploaded file\n",
        "        f.write(uploaded[next(iter(uploaded))])\n",
        "else:\n",
        "    # If no files were uploaded, execute the following block of code\n",
        "    print('No file uploaded.')\n",
        "\n",
        "    # Exit the program\n",
        "    sys.exit()\n",
        "\n",
        "\n",
        "\n",
        "#Detect ORFs in query\n",
        "print(\"Searching for RBP...\")\n",
        "for seq_record in SeqIO.parse(\"files/genome.fasta\", \"fasta\"):\n",
        "    orfseq = []\n",
        "    orflen = []\n",
        "    orfnr = []\n",
        "    x = 0\n",
        "    file = seq_record.seq\n",
        "    for strand, nuc in [(+1, file), (-1, file.reverse_complement())]:\n",
        "        for frame in range(3):\n",
        "            length = 3 * ((len(file)-frame) // 3) #Multiple of three\n",
        "            for pro in nuc[frame:frame+length].translate(11).split(\"*\"):\n",
        "                  splitlocal = pro.find('M')\n",
        "                  seq_final = pro[splitlocal:]\n",
        "                  if len(seq_final) >= 100:\n",
        "                    x = x+1\n",
        "                    orflen.append(len(seq_final))\n",
        "                    seq_final = str(seq_final)\n",
        "                    seq_final = '\\n'.join(seq_final[i:i+60] for i in range(0, len(seq_final), 60))\n",
        "                    orfseq.append(str(seq_final))\n",
        "                    orfnr.append(x)\n",
        "if len(orfseq) == 0:\n",
        "    print(\"No ORFs were detected.\")\n",
        "    sys.exit()\n",
        "else:\n",
        "    with open('files/query_db.fasta', 'w') as f:\n",
        "        for i in range(len(orfseq)):\n",
        "            f.write(\">\" + seq_record.id + \"_orf\" + str(orfnr[i]) + \" - length \" + str(orflen[i]) + \"\\n\")\n",
        "            f.write(orfseq[i] + \"\\n\")\n",
        "\n",
        "#Create database and blast with known RBPs\n",
        "!makeblastdb -in files/query_db.fasta -dbtype prot -out files/protein_db > /dev/null;\n",
        "for file in glob.glob(\"files/RBPs_main/*.fasta\"):\n",
        "        filename = file.replace(\"files/RBPs_main/\",\"\")\n",
        "        filename = filename.replace(\".fasta\",\"\")\n",
        "        !blastp -query {file} -db files/protein_db -outfmt 6 -out files/blast_main/{filename}_blast.txt\n",
        "for file in glob.glob(\"files/RBPs_rev/*.fasta\"):\n",
        "        filename = file.replace(\"files/RBPs_rev/\",\"\")\n",
        "        filename = filename.replace(\".fasta\",\"\")\n",
        "        !blastp -query {file} -db files/protein_db -outfmt 6 -out files/blast_rev/{filename}_blast.txt\n",
        "\n",
        "#Extract primary RBP from blast\n",
        "result = pd.DataFrame()\n",
        "for filename in glob.glob(\"files/blast_main/*_blast.txt\"):\n",
        "    df = pd.read_csv(filename, sep='\\t', header = None)\n",
        "    df.columns = ['query acc.ver', 'subject', 'identity', 'alignment length', 'mismatches', 'gap opens', 'q. start', 'q. end', 's. start', 's. end', 'evalue', 'bit score']\n",
        "    df[\"lengthper\"] = df[\"alignment length\"]/df['alignment length'].iloc[0]\n",
        "    df = df[df.lengthper >=0.8]\n",
        "    df = df[df.identity > 80]\n",
        "    df.loc[df['lengthper'] > 1, 'lengthper'] = 1\n",
        "    df = df.reset_index(drop=True)\n",
        "\n",
        "    if df.empty:\n",
        "      x = x\n",
        "    else:\n",
        "      result = pd.concat([result, df.sort_values('identity', ascending=False).head(1)], ignore_index=True)\n",
        "      finaldf = df\n",
        "\n",
        "# Graph settings\n",
        "x_labels = ['SA ΔtagO', 'SA WT', 'SA ΔtarM', 'SA ΔtarS', 'SA ΔtarM/S', 'SE 1457']\n",
        "colors = ['#F4FFC0', '#F4FFC0', '#F4FFC0', '#F4FFC0', '#F4FFC0', '#90BFF9']\n",
        "\n",
        "#Visualize primary RBP\n",
        "print(\"-----------------------------------------------------------------------------------------------------------\")\n",
        "RBPs_found = 0\n",
        "if result.empty:\n",
        "    print(\"No primary S. aureus binding phage RBP could be detected.\")\n",
        "else:\n",
        "    RBPs_found = 1\n",
        "    print(\"\\nPrimary S. aureus binding phage RBP detected\\n\")\n",
        "    for i in range(len(result)):\n",
        "      phageRBP = result.iloc[i][\"query acc.ver\"]\n",
        "      for name, lst in clusterlist_main.items():\n",
        "        for item in lst:\n",
        "            if phageRBP in item:\n",
        "                cluster = name\n",
        "      if \"phiK \" in cluster:\n",
        "        if \"subgroup a\" in cluster:\n",
        "          graphname = \"phiK\"\n",
        "        elif \"subgroup b\" in cluster:\n",
        "          graphname = \"phiStab20\"\n",
        "        elif \"subgroup c\" in cluster:\n",
        "          graphname = \"phiSA012\"\n",
        "        elif \"subgroup d\" in cluster:\n",
        "          graphname = \"phiPG2021-10\"\n",
        "        elif \"subgroup e\" in cluster:\n",
        "          graphname = \"phiPG2021-17\"\n",
        "        elif \"subgroup f\" in cluster:\n",
        "          graphname = \"not tested\"\n",
        "        elif \"outlier\" in cluster:\n",
        "          graphname = \"not tested\"\n",
        "      else:\n",
        "        graphname = cluster\n",
        "      if graphname in RBP_defi:\n",
        "        RBP_match = RBP_defi[graphname][0]\n",
        "        match_row = finaldf.loc[finaldf['query acc.ver'] == RBP_match]\n",
        "        identity_value = match_row.iloc[0]['identity']\n",
        "      print(\"-\\033[1mPRIMARY RBP \" + str(i+1) + \": \" + result.iloc[i][\"subject\"] + \"\\033[0m\")\n",
        "      print(\"-Clustered in group: \\033[1m\" + cluster + \"\\033[0m\")\n",
        "      print(\"-Highest sequence identity with RBP of phage \\033[4m\\033[1m\" + result.iloc[i][\"query acc.ver\"].split(\"_orf\")[0] + \"\\033[0m: \" + str(result.iloc[i][\"identity\"]) + \"%\")\n",
        "      print(\"-Overlap with main RBP of phage \" + result.iloc[i][\"query acc.ver\"].split(\"_orf\")[0] + \": \" + str(result.iloc[i][\"lengthper\"]*100) + \"%\")\n",
        "      print(\"-ORFs can be found in 'files/query_db.fasta'.\\n\")\n",
        "      with open('files/query_db.fasta', 'r') as f:\n",
        "          query_db = f.read()\n",
        "          query_db = query_db.split(result.iloc[0][\"subject\"])[-1]\n",
        "          query_db = query_db.split(\">\")[0]\n",
        "          print(\">\" + result.iloc[0][\"subject\"] + query_db + \"\\n\")\n",
        "      y_values = [0,0,0,0,0,0]\n",
        "      if graphname in FACS_values:\n",
        "          print(\"\\033[1mIdentity of RBP \" + result.iloc[i][\"subject\"] + \" with RBP of \" + graphname + \": \" + str(np.round(identity_value,2)) + \"%\\033[0m\")\n",
        "          y_values = FACS_values[graphname]\n",
        "          sem = FACS_SEM[graphname]\n",
        "          ind = np.arange(6)\n",
        "          # Create the bar chart with the desired size\n",
        "          fig, ax = plt.subplots(figsize=(3.5, 3.5))\n",
        "\n",
        "          # Create the bar plot\n",
        "          ax.bar(x_labels, y_values, align='center', alpha=1, ecolor=\"k\", color=colors, edgecolor=\"k\",\n",
        "                 linewidth=1.5)\n",
        "          plotline1, caplines1, barlinecols1 = ax.errorbar(ind, y_values, yerr=sem, lolims=True,\n",
        "                                                           capsize = 0, ls='None', color=\"k\", alpha=0.5,\n",
        "                                                           linewidth=1)\n",
        "\n",
        "          caplines1[0].set_marker('_')\n",
        "          caplines1[0].set_markersize(10)\n",
        "          # Set the axis labels and title\n",
        "          ax.set_ylim(bottom=0)\n",
        "          ax.set_xlabel('')\n",
        "          ax.set_ylabel('Median fluorescence intensity [AU]')\n",
        "          ax.set_title('Binding of ' + graphname + \" main RBP\")\n",
        "          ax.set_xticks(range(len(x_labels)))\n",
        "          ax.set_xticklabels(x_labels, rotation=45, ha='right')\n",
        "      else:\n",
        "          print(f\"No binding pattern available for {graphname}\")\n",
        "      plt.show()\n",
        "      print(f\"\\n-For more information, find the closest matching RBP under {'https://itol.embl.de/tree/4652552332411677924980'}\")\n",
        "if 'df' in locals():\n",
        "    del df\n",
        "if 'result' in locals():\n",
        "    del result\n",
        "\n",
        "#Extract secondary RBP from blast\n",
        "result = pd.DataFrame()\n",
        "for filename in glob.glob(\"files/blast_rev/*_blast.txt\"):\n",
        "    df = pd.read_csv(filename, sep='\\t', header = None)\n",
        "    df.columns = ['query acc.ver', 'subject', 'identity', 'alignment length', 'mismatches', 'gap opens', 'q. start', 'q. end', 's. start', 's. end', 'evalue', 'bit score']\n",
        "    df[\"lengthper\"] = df[\"alignment length\"]/df['alignment length'].iloc[0]\n",
        "    df = df[df.lengthper >=0.8]\n",
        "    df = df[df.identity > 80]\n",
        "    df.loc[df['lengthper'] > 1, 'lengthper'] = 1\n",
        "    df = df.reset_index(drop=True)\n",
        "    if df.empty:\n",
        "      x = x\n",
        "    else:\n",
        "      result = pd.concat([result, df.sort_values('identity', ascending=False).head(1)], ignore_index=True)\n",
        "      finaldf = df\n",
        "\n",
        "#Visualize secondary RBP\n",
        "print(\"-----------------------------------------------------------------------------------------------------------\")\n",
        "if result.empty:\n",
        "    print(\"No secondary S. aureus binding phage RBP could be detected.\")\n",
        "    if RBPs_found == 0:\n",
        "      print(\"-----------------------------------------------------------------------------------------------------------\")\n",
        "      print(\"\\033[1mNo RBPs found. This phage might not be able to infect S. aureus or is an outliers that does not cluster.\\033[0m\")\n",
        "else:\n",
        "    print(\"\\nSecondary S. aureus binding phage RBP detected\\n\")\n",
        "    for i in range(len(result)):\n",
        "      phageRBP = result.iloc[i][\"query acc.ver\"]\n",
        "      for name, lst in clusterlist_rev.items():\n",
        "        for item in lst:\n",
        "            if phageRBP in item:\n",
        "                cluster = name\n",
        "      if \"phiK_reversible \" in cluster:\n",
        "        if \"subgroup a\" in cluster:\n",
        "          graphname = \"phiK_reversible\"\n",
        "        elif \"subgroup b\" in cluster:\n",
        "          graphname = \"phiPG2021-10_reversible\"\n",
        "        elif \"outlier\" in cluster:\n",
        "          graphname = \"not tested\"\n",
        "      else:\n",
        "        graphname = cluster\n",
        "      if graphname in RBP_defi:\n",
        "        RBP_match = RBP_defi[graphname][0]\n",
        "        match_row = finaldf.loc[finaldf['query acc.ver'] == RBP_match]\n",
        "        identity_value = match_row.iloc[0]['identity']\n",
        "      print(\"-\\033[1mREVERSIBLE RBP \" + str(i+1) + \": \" + result.iloc[i][\"subject\"] + \"\\033[0m\")\n",
        "      print(\"-Clustered in group: \\033[1m\" + cluster + \"\\033[0m\")\n",
        "      print(\"-Highest sequence identity with RBP of phage \\033[4m\\033[1m\" + result.iloc[i][\"query acc.ver\"]\n",
        "            .split(\"_orf\")[0] + \"\\033[0m: \" + str(result.iloc[i][\"identity\"]) + \"%\")\n",
        "      print(\"-Overlap with main RBP of phage \" + result.iloc[i][\"query acc.ver\"].split(\"_orf\")[0] + \": \" + str(result.iloc[i][\"lengthper\"]*100) + \"%\")\n",
        "      print(\"-ORFs can be found in 'files/query_db.fasta'.\\n\")\n",
        "\n",
        "      with open('files/query_db.fasta', 'r') as f:\n",
        "          query_db = f.read()\n",
        "          query_db = query_db.split(result.iloc[0][\"subject\"])[-1]\n",
        "          query_db = query_db.split(\">\")[0]\n",
        "          print(\">\" + result.iloc[0][\"subject\"] + query_db + \"\\n\")\n",
        "      y_values = [0,0,0,0,0,0]\n",
        "      if graphname in FACS_values:\n",
        "          print(\"\\033[1mIdentity of RBP \" + result.iloc[i][\"subject\"] + \" with RBP of \" + graphname + \":\" + str(np.round(identity_value,2)) + \"%\\033[0m\")\n",
        "          y_values = FACS_values[graphname]\n",
        "          sem = FACS_SEM[graphname]\n",
        "          ind = np.arange(6)\n",
        "          # Create the bar chart with the desired size\n",
        "          fig, ax = plt.subplots(figsize=(3.5, 3.5))\n",
        "\n",
        "          # Create the bar plot\n",
        "          ax.bar(x_labels, y_values, align='center', alpha=1, ecolor=\"k\", color=colors, edgecolor=\"k\",\n",
        "                 linewidth=1.5)\n",
        "          plotline1, caplines1, barlinecols1 = ax.errorbar(ind, y_values, yerr=sem, lolims=True,\n",
        "                                                           capsize = 0, ls='None', color=\"k\", alpha=0.5,\n",
        "                                                           linewidth=1)\n",
        "\n",
        "          caplines1[0].set_marker('_')\n",
        "          caplines1[0].set_markersize(10)\n",
        "          # Set the axis labels and title\n",
        "          ax.set_ylim(bottom=0)\n",
        "          ax.set_xlabel('')\n",
        "          ax.set_ylabel('Median fluorescence intensity [AU]')\n",
        "          ax.set_title('Binding of ' + graphname + \" RBP\")\n",
        "          ax.set_xticks(range(len(x_labels)))\n",
        "          ax.set_xticklabels(x_labels, rotation=45, ha='right')\n",
        "      else:\n",
        "          print(f\"No binding pattern available for {graphname}\")\n",
        "      plt.show()\n",
        "      print(f\"\\n-For more information, find the closest matching RBP under {'https://itol.embl.de/tree/46525523312761688382866'}\")"
      ]
    }
  ]
}
